### PR TITLE
Feature/get weather

### PR DESCRIPTION
**Pull request by ChristianBorn**

### CODE REVIEW

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherService.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        if (hour < 0 || hour > 24) {
+            throw new IllegalArgumentException("Hour must be between 0 and 24");
+        }
+        WeatherForecastResponse weatherResponse = Objects.requireNonNull(webClient
+                        .get()
+                        .uri("?wmo_station_id=10637&date=" + date)
+                        .retrieve()
+                        .toEntity(WeatherForecastResponse.class)
+                        .block())
+                        .getBody();
```
Comment (Kreyno93) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherService.java:29]: Try to seperate the response from the body. 

` ResponseEntity<WeatherForecastResponse> weatherResponse = (webClient
                        .get()
                        .uri("?wmo_station_id=10637&date=" + date)
                        .retrieve()
                        .toEntity(WeatherForecastResponse.class)
                        .block());`

File: frontend/src/Weather/WeatherWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    const [weather, setWeather] = useState<WeatherModel>()
+    const [date, setDate] = useState<Date>(new Date())
+
+    const fetchWeather = () => {
+        const urlDate = getDate(date)
+        const hour = getTime(date)
+        axios.get("/api/weather/"+urlDate+"?hour="+hour)
+            .then(response => response.data)
+            .catch(error => console.log(error))
```
Comment (Kreyno93) [inline frontend/src/Weather/WeatherWidget.tsx:18]: itll be better to use console.error instead of console.log

**Review by Kreyno93 — CHANGES_REQUESTED** [review]

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherController.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+public class WeatherController {
+
+    private final WeatherService weatherService;
+
+    public WeatherController(WeatherService weatherService) {
+        this.weatherService = weatherService;
+    }
+
+    @GetMapping("/{date}")
+    public WeatherData getWeather(@PathVariable String date, @RequestParam int hour) {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherController.java:19]: I would probalby not transmit the date as a Pathvariable and the time as a requestparameter... Im not sure how your api works. But maybe you could send it combined as a Date-Time Object (Instant?) or maybe seperated as two Request Parameters...

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package de.neuefische.ffmjava221.teamprojekt.backend.weather;
+
+import java.util.List;
+
+public record WeatherForecastResponse(List<WeatherData> weather) {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java:5]: I dont like the namings of your two record and their attributes.
Maybe you could name this record `DailyWeather` with the List `hourlyWeather`? 🤔

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherService.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+@Service
+public class WeatherService {
+    private final WebClient webClient;
+
+    public WeatherService(@Value("${baseUrl}") String baseUrl) {
+        this.webClient = WebClient.create(baseUrl);
+    }
+
+    public WeatherData fetchWeather(String date, int hour) throws IllegalArgumentException, WeatherResponseIsNullException {
+        if (hour < 0 || hour > 24) {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherService.java:20]: Not all days have 24 hours... ⏲️

File: backend/src/main/resources/application.properties [hunk 6]
```diff
@@ -1 +1,2 @@
+baseUrl=https://api.brightsky.dev/weather
```
Comment (Flooooooooooorian) [inline backend/src/main/resources/application.properties:1]:
```suggestion
weather.api.baseUrl=https://api.brightsky.dev/weather
```

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class WeatherIntegrationTest {
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherIntegrationTest.java:16]: We want to mock external apis in our test, like you did in your other test class. Either remove this one or add your mockwebserver

File: backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherMockWebServerIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        registry.add("baseUrl", () -> mockWebServer.url("/").toString());
+    }
+
+    @AfterAll
+    static void afterAll() throws IOException {
+        mockWebServer.shutdown();
+    }
+
+    @Test
+    void fetchWeatherDateNullReturnsError() throws Exception {
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherMockWebServerIntegrationTest.java:45]: nice 👍 
I would add at least one positive test case

File: frontend/src/Weather/WeatherWidget.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        setDate(date)
+    }
+
+    const handleDateSelect = () => {
+        fetchWeather()
+    }
+
+    return (
+        <div>
+            {weather !== undefined ?
```
Comment (Flooooooooooorian) [inline frontend/src/Weather/WeatherWidget.tsx:44]:
```suggestion
            {weather ?
```

File: backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java (hunk 4 above)
Comment (Dskottke) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java:5]: Isn't it a problem that the name of the List have to match with the list name in the JSON ?

Comment (Flooooooooooorian) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java:5]: Yeah it is. But you can use  the `JsonProperty ` Annotation to change the name for the serialization/deserialization

Comment (Dskottke) [reply backend/src/main/java/de/neuefische/ffmjava221/teamprojekt/backend/weather/WeatherForecastResponse.java:5]: Ahh cool thank u!!!

**Review by Flooooooooooorian — DISMISSED** [review]

**Review by Kreyno93 — DISMISSED** [review]

**Review by Kreyno93 — APPROVED** [review]

**Pull request merged**